include/pybind11/stl.h

```diff
@@ -49,6 +49,13 @@
 NAMESPACE_BEGIN(pybind11)
 NAMESPACE_BEGIN(detail)
 
+// Extracts an const lvalue reference or rvalue reference for a container element based on the
+// lvalue/rvalue status of the container
+template <typename ContainerArg, typename Value>
+using forward_item = conditional_t<
+    std::is_lvalue_reference<ContainerArg>::value, const Value &,
+    typename std::add_rvalue_reference<Value>::type>;
```

Comment:
Explicitly using `const Value &` is a bit less generic than it could be: `remove_reference_t<Value>&` and `remove_reference_t<Value>&&`.

Consider wrapping it in a `forward_like` function like here: https://stackoverflow.com/a/29780197/7189436
`forward_like<Container>(element)` is pretty nice and expressive.